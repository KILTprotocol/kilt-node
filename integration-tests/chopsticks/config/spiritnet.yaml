endpoint:
 - wss://kilt-rpc.dwellir.com
db: ./db/spiritnet.db.sqlite
port: ${env.SPIRITNET_PORT}
# wasm-override: ../../target/debug/wbuild/spiritnet-runtime/spiritnet_runtime.wasm
runtime-log-level: 5

import-storage:
 TechnicalCommittee:
  Members:
   - 4rsmbFBYpVmWE2LHRsSZKxf22a8cbJLxMZqvVGkGkSDmDBcr
 Council:
  Members:
   - 4rsmbFBYpVmWE2LHRsSZKxf22a8cbJLxMZqvVGkGkSDmDBcr
 System:
  Account:
   -
    -
     - 4rsmbFBYpVmWE2LHRsSZKxf22a8cbJLxMZqvVGkGkSDmDBcr
    - providers: 1
      data:
       free: "1000000000000000000000"
   -
    -
     - 4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ
    - providers: 1
      data:
       free: "1000000000000000000000"
