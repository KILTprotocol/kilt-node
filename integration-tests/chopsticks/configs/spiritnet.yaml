endpoint: wss://spiritnet.kilt.io
# mock-signature-host: true
db: out/spiritnet/db.sqlite
