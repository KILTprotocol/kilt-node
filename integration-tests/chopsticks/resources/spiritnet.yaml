endpoint: wss://spiritnet.kilt.io
mock-signature-host: true
db: ./db/spiritnet.db.sqlite
port: 8002

import-storage:
    technicalCommittee:
        Members: ["4pF5Y2Eo6doQHPLQj5AkndZwtomVB8ab2sRftRS2D9JDdELr"]
    council:
        Members: ["4pF5Y2Eo6doQHPLQj5AkndZwtomVB8ab2sRftRS2D9JDdELr"]
