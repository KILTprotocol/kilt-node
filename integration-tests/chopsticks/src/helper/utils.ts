import { ApiPromise, SubmittableResult } from '@polkadot/api'
import { Keyring } from '@polkadot/keyring'
import { u8aToHex } from '@polkadot/util'
import { decodeAddress } from '@polkadot/util-crypto'

import type { SubmittableExtrinsic } from '@polkadot/api/types'
import type { KeyringPair } from '@polkadot/keyring/types'
import type { Codec } from '@polkadot/types/types'
import type { ExpectStatic } from 'vitest'
import { EventRecord } from '@polkadot/types/interfaces'

/**
 * Keyring instance to generate key pairs with 38 ss58Format encoding.
 */
const keyring = new Keyring({ type: 'ed25519', ss58Format: 38 })

/**
 * Key pairs for Alice, Bob, and Charlie used for testing.
 */
export const keysAlice = keyring.addFromUri('//alice', undefined)
export const keysBob = keyring.addFromUri('//bob', undefined)
export const keysCharlie = keyring.addFromUri('//charlie', undefined)

/**
 * Converts a given value to a Number
 */
export function toNumber(value: string | undefined): number | undefined {
	if (value === undefined) {
		return undefined
	}
	return Number(value)
}

/**
 * Converts a given address to its hexadecimal representation.
 *
 * @param addr - The address to be converted. It should be a string representation of an address.
 */
export function hexAddress(addr: string) {
	return u8aToHex(decodeAddress(addr))
}

/**
 * Validates if the received balance is within the expected range considering a certain precision.
 *
 * @param previousBalance - The balance before the operation.
 * @param receivedBalance - The balance after the operation.
 * @param deltaBalance - The expected change in balance.
 * @param expect - The assertion function from the testing library.
 * @param precision - The precision of the balance validation. It must be a value between 0 and 100.
 *
 * The function calculates the expected balance by adding the deltaBalance to the previousBalance.
 * It then calculates the lower and upper bounds of the expected balance considering the precision.
 * The received balance is then checked if it falls within the range of the lower and upper bounds.
 *
 * If the precision is not between 0 and 100, the function throws an error.
 *
 * @throws {Error} If the precision is not between 0 and 100.
 */
export function validateBalanceWithPrecision(
	previousBalance: bigint,
	receivedBalance: bigint,
	deltaBalance: bigint,
	expect: ExpectStatic,
	precision: bigint
) {
	if (precision < 0n || precision > 100n) {
		throw new Error('Precision must be between 0 and 100')
	}

	const allowedError = 100n - precision
	const expectedBalance = previousBalance + deltaBalance

	const lowerBound = expectedBalance - (expectedBalance * allowedError) / 100n
	const upperBound = expectedBalance + (expectedBalance * allowedError) / 100n

	expect(receivedBalance).toBeGreaterThanOrEqual(lowerBound)
	expect(receivedBalance).toBeLessThanOrEqual(upperBound)
}

/**
 * Fetches the paid fees for the executed XCM message. Is only working on the sender chain.
 */
export async function getPaidXcmFees(api: ApiPromise, eventRecords: Codec[]): Promise<bigint> {
	const polkadotFees = eventRecords.find((eventRecord) =>
		api.events.polkadotXcm.FeesPaid.is((eventRecord as EventRecord).event)
	)
	if (!polkadotFees) {
		throw new Error('FeesPaid event not found')
	}

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	return (polkadotFees as any).event.data.fees[0].fun.asFungible.toBigInt()
}

/**
 * Calculates the transaction fees for a given transaction.
 */
export async function calculateTxFees(
	tx: SubmittableExtrinsic<'promise', SubmittableResult>,
	account: KeyringPair
): Promise<bigint> {
	const paymentInfo = await tx.paymentInfo(account)
	return paymentInfo.partialFee.toBigInt()
}

// Constants for the balances
export const KILT = BigInt(1e15)
export const DOT = BigInt(1e10)
export const HDX = BigInt(1e12)

export const initialBalanceKILT = 100n * KILT
export const initialBalanceDOT = 100n * DOT
export const initialBalanceHDX = 100n * HDX
