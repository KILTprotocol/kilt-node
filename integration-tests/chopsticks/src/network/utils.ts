export function getSiblingLocation(paraId: number) {
	return {
		parents: 1,
		interior: {
			X1: { Parachain: paraId },
		},
	}
}

export function getParentLocation() {
	return {
		parents: 1,
		interior: 'Here',
	}
}

export function getAccountLocationV2(addr: string) {
	return {
		V2: {
			parents: 0,
			interior: {
				X1: {
					AccountId32: {
						network: 'Any',
						id: addr,
					},
				},
			},
		},
	}
}

export function getAccountLocationV3(addr: string) {
	return {
		V3: {
			parents: 0,
			interior: {
				X1: {
					AccountId32: {
						id: addr,
					},
				},
			},
		},
	}
}

export function getNativeAssetIdLocation(amount: bigint) {
	return {
		id: { Concrete: { parents: 0, interior: 'Here' } },
		fun: { Fungible: amount },
	}
}
