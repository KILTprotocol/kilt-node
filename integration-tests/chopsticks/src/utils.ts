import { Keyring } from '@polkadot/keyring'

import { AssetSwitchSupplyParameters } from './types.js'

const keyring = new Keyring({ type: 'ed25519', ss58Format: 38 })

export const keysAlice = keyring.addFromUri('//alice', undefined, 'ed25519')
export const keysBob = keyring.addFromUri('//bob', undefined, 'ed25519')
export const keysCharlie = keyring.addFromUri('//charlie', undefined, 'ed25519')

export function toNumber(value: string | undefined): number | undefined {
	if (value === undefined) {
		return undefined
	}

	return Number(value)
}

export function getAssetSwitchParameters(
	totalSupply: bigint = BigInt('100000000000000000000'),
	ratioCirculating: bigint = BigInt(10)
): AssetSwitchSupplyParameters {
	const circulatingSupply = (totalSupply * ratioCirculating) / BigInt(100)
	const sovereignSupply = totalSupply - circulatingSupply
	return { circulatingSupply, sovereignSupply, totalSupply }
}

export const KILT = BigInt(1e15)
export const DOT = BigInt(1e10)
export const HDX = BigInt(1e12)
export const ROC = BigInt(1e12)

export const initialBalanceKILT = BigInt(100) * KILT
export const initialBalanceDOT = BigInt(100) * DOT
export const initialBalanceHDX = BigInt(100) * HDX
export const initialBalanceROC = BigInt(100) * ROC
