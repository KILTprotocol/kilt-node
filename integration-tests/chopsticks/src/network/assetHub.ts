import { setupContext, SetupOption } from '@acala-network/chopsticks-testing'

import type { Config } from './types.js'
import { initialBalanceDOT, toNumber } from '../utils.js'

/// Options used to create the Spiritnet context
export const getSetupOptions = ({
	blockNumber = undefined,
	wasmOverride = undefined,
}: {
	blockNumber?: number
	wasmOverride?: string
}) =>
	({
		endpoint: process.env.ASSETHUB_WSS || 'wss://asset-hub-polkadot-rpc.dwellir.com',
		db: './db/assethub.db.sqlite',
		port: toNumber(process.env.ASSETHUB_PORT) || 9003,
		wasmOverride,
		blockNumber,
	}) as SetupOption

/// AssetHub has no own coin. Teleported dots are used as the native token.
export function assignDotTokensToAccountsAsStorage(addr: string[], balance: bigint = initialBalanceDOT) {
	return {
		System: {
			Account: addr.map((address) => [[address], { providers: 1, data: { free: balance.toString() } }]),
		},
	}
}

export function createForeignAsset(manager: string, assetId = eKiltLocation) {
	return {
		foreignAssets: {
			asset: [
				[
					[assetId],
					{
						owner: manager,
						issuer: manager,
						admin: manager,
						freezer: manager,
						// Just make it big enough
						supply: '10000000000000000000000000000',
						deposit: 0,
						minBalance: 0,
						isSufficient: false,
						accounts: 0,
						sufficients: 0,
						approvals: 0,
						status: 'Live',
					},
				],
			],
		},
	}
}

/// Assigns KSM to an account
export function assignKSMtoAccounts(addr: string[], balance: bigint = initialBalanceDOT) {
	return {
		foreignAssets: {
			account: addr.map((addr) => [
				[KSMAssetLocation, addr],
				{
					balance: balance,
					status: 'Liquid',
					reason: 'Consumer',
					extra: null,
				},
			]),
		},
	}
}

/// Assigns the foreign asset to the accounts.
/// Does not check if supply is matching the sum of the account balances.
export function assignForeignAssetToAccounts(accountInfo: [string, bigint][], assetId = eKiltLocation) {
	return {
		foreignAssets: {
			account: accountInfo.map(([account, balance]) => [
				[assetId, account],
				{
					balance: balance,
					status: 'Liquid',
					reason: 'Consumer',
					extra: null,
				},
			]),
		},
	}
}

/// AssetHub ParaId
export const paraId = 1000

export const KSMAssetLocation = {
	parents: 2,
	interior: {
		X1: {
			GlobalConsensus: 'Kusama',
		},
	},
}

// Sibling Sovereign Account
export const sovereignAccountOnSiblingChains = '4qXPdpimHh8TR24RSk994yVzxx4TLfvKj5i1qH5puvWmfAqy'

/// Native token in AssetHub
export const nativeTokenLocation = { parents: 1, interior: 'Here' }

export const eKiltLocation = {
	parents: 2,
	interior: {
		X2: [
			{
				GlobalConsensus: { Ethereum: { chainId: 11155111 } },
			},
			{
				AccountKey20: {
					network: null,
					key: '0x06012c8cf97bead5deae237070f9587f8e7a266d',
				},
			},
		],
	},
}

export async function getContext(): Promise<Config> {
	const options = getSetupOptions({})
	return setupContext(options)
}
