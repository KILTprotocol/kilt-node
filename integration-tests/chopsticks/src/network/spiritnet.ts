import { setupContext, SetupOption } from '@acala-network/chopsticks-testing'
import type { Config } from './types.js'
import { initialBalanceKILT, toNumber } from '../utils.js'

/// Options used to create the Spiritnet context
const options: SetupOption = {
	endpoint: process.env.SPIRITNET_WS || 'wss://kilt-rpc.dwellir.com',
	db: './db/spiritnet.db.sqlite',
	port: toNumber(process.env.SPIRITNET_PORT) || 9002,
}

/// Assigns the native tokens to an accounts
export function assignNativeTokensToAccounts(addr: string[], balance: bigint = initialBalanceKILT) {
	return {
		System: {
			Account: addr.map((address) => [[address], { providers: 1, data: { free: balance } }]),
		},
	}
}

/// Sets the [technicalCommittee] and [council] governance to the given accounts
export function setGovernance(addr: string[]) {
	return {
		technicalCommittee: { Members: addr },
		council: { Members: addr },
	}
}

export function XcmPalletSafeVersion3StorageEntry() {
	return {
		polkadotXcm: {
			safeXcmVersion: 3,
		},
	}
}

/// Spiritnet ParaId
export const paraId = 2086

/// The sovereign account of HydraDx in Spiritnet
export const hydraDxSovereignAccount = '4qXPdpioJ6D8cgdeYXaukV2Y2oAQUHaX1VnGhdbSRqJn2CBt'

/// Returns the Spiritnet context for the given options
export async function getContext(): Promise<Config> {
	return setupContext(options)
}
