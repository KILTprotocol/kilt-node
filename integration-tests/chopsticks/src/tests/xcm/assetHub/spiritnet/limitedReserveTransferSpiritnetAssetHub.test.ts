import { test } from 'vitest'
import { sendTransaction, withExpect } from '@acala-network/chopsticks-testing'

import * as SpiritnetConfig from '../../../../network/spiritnet.js'
import * as AssetHubConfig from '../../../../network/assetHub.js'
import { KILT, initialBalanceKILT, keysAlice } from '../../../../utils.js'
import { spiritnetContext, assethubContext, getFreeBalanceSpiritnet } from '../../../index.js'
import { getAccountLocationV3, getNativeAssetIdLocationV3, getSiblingLocationV3 } from '../../../../network/utils.js'
import { checkBalance, createBlock, hexAddress, setStorage } from '../../../utils.js'

test('Limited Reserve Transfers from Spiritnet Account Alice -> AH Account Alice', async ({ expect }) => {
	const { checkEvents, checkSystemEvents } = withExpect(expect)

	// Assign alice some KILT tokens
	await setStorage(spiritnetContext, {
		...SpiritnetConfig.assignNativeTokensToAccounts([keysAlice.address], initialBalanceKILT),
		...SpiritnetConfig.XcmPalletSafeVersion3StorageEntry(),
	})

	// Balance of the AH sovereign account before the transfer
	const assetHubSovereignAccountBalance = await getFreeBalanceSpiritnet(
		AssetHubConfig.sovereignAccountOnSiblingChains
	)

	// check initial balance of Alice on Spiritnet
	await checkBalance(getFreeBalanceSpiritnet, keysAlice.address, expect, initialBalanceKILT)

	const aliceAddress = hexAddress(keysAlice.address)
	const assetHubDestination = { V3: getSiblingLocationV3(AssetHubConfig.paraId) }
	const beneficiary = getAccountLocationV3(aliceAddress)
	const kiltAsset = { V3: [getNativeAssetIdLocationV3(KILT.toString())] }

	const signedTx = spiritnetContext.api.tx.polkadotXcm
		.limitedReserveTransferAssets(assetHubDestination, beneficiary, kiltAsset, 0, 'Unlimited')
		.signAsync(keysAlice)

	const events = await sendTransaction(signedTx)

	// Check sender state
	await createBlock(spiritnetContext)

	// Check events sender
	await checkEvents(events, 'xcmpQueue').toMatchSnapshot(
		`sender spiritnet::xcmpQueue::[XcmpMessageSent] asset ${JSON.stringify(kiltAsset)}`
	)
	await checkEvents(events, 'polkadotXcm').toMatchSnapshot(
		`sender spiritnet::polkadotXcm::[Attempted] asset ${JSON.stringify(kiltAsset)}`
	)
	await checkEvents(events, { section: 'balances', method: 'Withdraw' }).toMatchSnapshot(
		`sender spiritnet::balances::[Withdraw] asset ${JSON.stringify(kiltAsset)}`
	)

	//	check balance. The sovereign account should hold one additional KILT.
	await checkBalance(
		getFreeBalanceSpiritnet,
		AssetHubConfig.sovereignAccountOnSiblingChains,
		expect,
		assetHubSovereignAccountBalance + KILT
	)

	await createBlock(assethubContext)

	// MSG processing will fail on AH.
	await checkSystemEvents(assethubContext, 'messageQueue').toMatchSnapshot(
		`receiver assetHub::messageQueue::[Processed] asset ${JSON.stringify(kiltAsset)}`
	)
}, 20_000)
