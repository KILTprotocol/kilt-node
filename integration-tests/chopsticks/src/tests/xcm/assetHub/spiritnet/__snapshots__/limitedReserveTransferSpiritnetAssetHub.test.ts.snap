// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Limited Reserve Transfers from Spiritnet Account Alice -> AH Account Alice > receiver assetHub::messageQueue::[Processed] asset {"V3":[{"id":{"Concrete":{"parents":0,"interior":"Here"}},"fun":{"Fungible":"1000000000000000"}}]} 1`] = `
[
  {
    "data": {
      "id": "(hash)",
      "origin": {
        "Sibling": "(rounded 2100)",
      },
      "success": false,
      "weightUsed": {
        "proofSize": "(rounded 1500)",
        "refTime": "(rounded 29000000)",
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;

exports[`Limited Reserve Transfers from Spiritnet Account Alice -> AH Account Alice > sender spiritnet::balances::[Withdraw] asset {"V3":[{"id":{"Concrete":{"parents":0,"interior":"Here"}},"fun":{"Fungible":"1000000000000000"}}]} 1`] = `
[
  {
    "data": {
      "amount": "(rounded 170000000000)",
      "who": "4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ",
    },
    "method": "Withdraw",
    "section": "balances",
  },
]
`;

exports[`Limited Reserve Transfers from Spiritnet Account Alice -> AH Account Alice > sender spiritnet::polkadotXcm::[Attempted] asset {"V3":[{"id":{"Concrete":{"parents":0,"interior":"Here"}},"fun":{"Fungible":"1000000000000000"}}]} 1`] = `
[
  {
    "data": {
      "outcome": {
        "Complete": {
          "proofSize": 0,
          "refTime": 400000000,
        },
      },
    },
    "method": "Attempted",
    "section": "polkadotXcm",
  },
]
`;

exports[`Limited Reserve Transfers from Spiritnet Account Alice -> AH Account Alice > sender spiritnet::xcmpQueue::[XcmpMessageSent] asset {"V3":[{"id":{"Concrete":{"parents":0,"interior":"Here"}},"fun":{"Fungible":"1000000000000000"}}]} 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
    },
    "method": "XcmpMessageSent",
    "section": "xcmpQueue",
  },
]
`;
